### ISSUE TITLE
qryn Kubernetes setup initialization error

### ISSUE THREAD

**Issue by JonnyBDev** [metrico/qryn#349]
Hello,

we are currently trying to deploy qryn to our Kubernetes cluster. For Click we are using the latest Bitnami chart (https://artifacthub.io/packages/helm/bitnami/clickhouse). We are deploying Clickhouse (23.8.2) in default settings with a username and password. We are deploying three replicas - everything in Clickhouse is looking fine. It's important to mention that we are running the LATEST available version of Clickhouse and not the version you recommend (ClickHouse Server LTS version >22 and <=22.9 is suggested). Using a recommended version would force us to use a chart that's from November last year.

![image](https://github.com/metrico/qryn/assets/20868670/f07fbc2c-5403-454a-a12f-69a40aae467d)

We are deploying qryn in the latest version too. On startup we are getting the following error log:
```
> qryn@2.3.12 start
> node qryn.js

{"level":30,"time":1695812755816,"pid":19,"hostname":"qryn-654c6c5db7-btgr6","name":"qryn","msg":"Initializing DB... cloki"}
{"level":30,"time":1695812755853,"pid":19,"hostname":"qryn-654c6c5db7-btgr6","name":"qryn","msg":"xxh ready"}
{"level":30,"time":1695812756120,"pid":19,"hostname":"qryn-654c6c5db7-btgr6","name":"qryn","msg":"v0 -> v1"}
{"level":20,"time":1695812756120,"pid":19,"hostname":"qryn-654c6c5db7-btgr6","name":"qryn","msg":"CREATE TABLE IF NOT EXISTS time_series (date Date,fingerprint UInt64,labels String, name String)\n    ENGINE = ReplacingMergeTree(date) PARTITION BY date ORDER BY fingerprint"}
{"level":30,"time":1695812756240,"pid":19,"hostname":"qryn-654c6c5db7-btgr6","name":"qryn","msg":"xxh ready"}
{"level":50,"time":1695812756311,"pid":19,"hostname":"qryn-654c6c5db7-btgr6","name":"qryn","err":"Error: Request failed with status code 404\nResponse: [404] undefined\nError: Request failed with status code 404\n    at createError (/app/node_modules/axios/lib/core/createError.js:16:15)\n    at settle (/app/node_modules/axios/lib/core/settle.js:17:12)\n    at IncomingMessage.handleStreamEnd (/app/node_modules/axios/lib/adapters/http.js:269:11)\n    at IncomingMessage.emit (node:events:525:35)\n    at endReadableNT (node:internal/streams/readable:1358:12)\n    at processTicksAndRejections (node:internal/process/task_queues:83:21)","msg":"Error starting qryn"}
npm notice 
npm notice New major version of npm available! 8.19.4 -> 10.1.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v10.1.0>
npm notice Run `npm install -g npm@10.1.0` to update!
npm notice
```

The Clickhouse server is giving us the following logs:
```
8. Poco::Net::TCPServerDispatcher::run() @ 0x0000000015b116d1 in /opt/bitnami/clickhouse/bin/clickhouse
9. Poco::PooledThread::run() @ 0x0000000015c47f07 in /opt/bitnami/clickhouse/bin/clickhouse
10. Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000015c461dc in /opt/bitnami/clickhouse/bin/clickhouse
11. start_thread @ 0x0000000000007ea7 in /lib/x86_64-linux-gnu/libpthread-2.31.so
12. ? @ 0x00000000000fba2f in /lib/x86_64-linux-gnu/libc-2.31.so
 (version 23.8.2.7 (official build))
2023.09.27 10:57:50.029193 [ 41 ] {} <Error> DynamicQueryHandler: Code: 81. DB::Exception: Database cloki does not exist. (UNKNOWN_DATABASE), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c604bf7 in /opt/bitnami/clickhouse/bin/clickhouse
1. DB::Exception::Exception<String>(int, FormatStringHelperImpl<std::type_identity<String>::type>, String&&) @ 0x000000000711518d in /opt/bitnami/clickhouse/bin/clickhouse
2. DB::DatabaseCatalog::assertDatabaseExists(String const&) const @ 0x00000000116cb88f in /opt/bitnami/clickhouse/bin/clickhouse
3. DB::Context::setCurrentDatabase(String const&) @ 0x0000000011610c31 in /opt/bitnami/clickhouse/bin/clickhouse
4. DB::HTTPHandler::processQuery(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&, DB::HTTPHandler::Output&, std::optional<DB::CurrentThread::QueryScope>&) @ 0x00000000130b26ec in /opt/bitnami/clickhouse/bin/clickhouse
5. DB::HTTPHandler::handleRequest(DB::HTTPServerRequest&, DB::HTTPServerResponse&) @ 0x00000000130b77a9 in /opt/bitnami/clickhouse/bin/clickhouse
6. DB::HTTPServerConnection::run() @ 0x0000000013128152 in /opt/bitnami/clickhouse/bin/clickhouse
7. Poco::Net::TCPServerConnection::start() @ 0x0000000015b104d4 in /opt/bitnami/clickhouse/bin/clickhouse
8. Poco::Net::TCPServerDispatcher::run() @ 0x0000000015b116d1 in /opt/bitnami/clickhouse/bin/clickhouse
9. Poco::PooledThread::run() @ 0x0000000015c47f07 in /opt/bitnami/clickhouse/bin/clickhouse
10. Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000015c461dc in /opt/bitnami/clickhouse/bin/clickhouse
11. start_thread @ 0x0000000000007ea7 in /lib/x86_64-linux-gnu/libpthread-2.31.so
12. ? @ 0x00000000000fba2f in /lib/x86_64-linux-gnu/libc-2.31.so
 (version 23.8.2.7 (official build))
2023.09.27 11:00:42.598889 [ 42 ] {} <Error> DynamicQueryHandler: Code: 81. DB::Exception: Database cloki does not exist. (UNKNOWN_DATABASE), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c604bf7 in /opt/bitnami/clickhouse/bin/clickhouse
1. DB::Exception::Exception<String>(int, FormatStringHelperImpl<std::type_identity<String>::type>, String&&) @ 0x000000000711518d in /opt/bitnami/clickhouse/bin/clickhouse
2. DB::DatabaseCatalog::assertDatabaseExists(String const&) const @ 0x00000000116cb88f in /opt/bitnami/clickhouse/bin/clickhouse
3. DB::Context::setCurrentDatabase(String const&) @ 0x0000000011610c31 in /opt/bitnami/clickhouse/bin/clickhouse
4. DB::HTTPHandler::processQuery(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&, DB::HTTPHandler::Output&, std::optional<DB::CurrentThread::QueryScope>&) @ 0x00000000130b26ec in /opt/bitnami/clickhouse/bin/clickhouse
5. DB::HTTPHandler::handleRequest(DB::HTTPServerRequest&, DB::HTTPServerResponse&) @ 0x00000000130b77a9 in /opt/bitnami/clickhouse/bin/clickhouse
6. DB::HTTPServerConnection::run() @ 0x0000000013128152 in /opt/bitnami/clickhouse/bin/clickhouse
7. Poco::Net::TCPServerConnection::start() @ 0x0000000015b104d4 in /opt/bitnami/clickhouse/bin/clickhouse
8. Poco::Net::TCPServerDispatcher::run() @ 0x0000000015b116d1 in /opt/bitnami/clickhouse/bin/clickhouse
9. Poco::PooledThread::run() @ 0x0000000015c47f07 in /opt/bitnami/clickhouse/bin/clickhouse
10. Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000015c461dc in /opt/bitnami/clickhouse/bin/clickhouse
11. start_thread @ 0x0000000000007ea7 in /lib/x86_64-linux-gnu/libpthread-2.31.so
12. ? @ 0x00000000000fba2f in /lib/x86_64-linux-gnu/libc-2.31.so
 (version 23.8.2.7 (official build))
2023.09.27 11:05:56.298589 [ 42 ] {} <Error> DynamicQueryHandler: Code: 81. DB::Exception: Database cloki does not exist. (UNKNOWN_DATABASE), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c604bf7 in /opt/bitnami/clickhouse/bin/clickhouse
1. DB::Exception::Exception<String>(int, FormatStringHelperImpl<std::type_identity<String>::type>, String&&) @ 0x000000000711518d in /opt/bitnami/clickhouse/bin/clickhouse
2. DB::DatabaseCatalog::assertDatabaseExists(String const&) const @ 0x00000000116cb88f in /opt/bitnami/clickhouse/bin/clickhouse
3. DB::Context::setCurrentDatabase(String const&) @ 0x0000000011610c31 in /opt/bitnami/clickhouse/bin/clickhouse
4. DB::HTTPHandler::processQuery(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&, DB::HTTPHandler::Output&, std::optional<DB::CurrentThread::QueryScope>&) @ 0x00000000130b26ec in /opt/bitnami/clickhouse/bin/clickhouse
5. DB::HTTPHandler::handleRequest(DB::HTTPServerRequest&, DB::HTTPServerResponse&) @ 0x00000000130b77a9 in /opt/bitnami/clickhouse/bin/clickhouse
6. DB::HTTPServerConnection::run() @ 0x0000000013128152 in /opt/bitnami/clickhouse/bin/clickhouse
7. Poco::Net::TCPServerConnection::start() @ 0x0000000015b104d4 in /opt/bitnami/clickhouse/bin/clickhouse
8. Poco::Net::TCPServerDispatcher::run() @ 0x0000000015b116d1 in /opt/bitnami/clickhouse/bin/clickhouse
9. Poco::PooledThread::run() @ 0x0000000015c47f07 in /opt/bitnami/clickhouse/bin/clickhouse
10. Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000015c461dc in /opt/bitnami/clickhouse/bin/clickhouse
11. start_thread @ 0x0000000000007ea7 in /lib/x86_64-linux-gnu/libpthread-2.31.so
12. ? @ 0x00000000000fba2f in /lib/x86_64-linux-gnu/libc-2.31.so
 (version 23.8.2.7 (official build))
```

I think, as @lmangani mentioned in Matrix that this belongs to a change in the API structure for Clickhouse. Would be very cool if we could figure out a fix for this problem. I am available @ Matrix too for quick response. If everything works fine in the end we could provide you with our chart that you have an "official" deployment for k8s.

If you need anything else feel free to reach out to us.

Thanks!

**Comment by JonnyBDev** [comment]
Additionally it's important to mention that I've tried it with Chart version 1.0.4 (Clickhouse 22.9.5) and it successfully initialized Clickhouse. So it's obvious that the recommended version is working but not the "latest.

**Comment by lmangani** [comment]
@JonnyBDev the 22.9.5 success seems curious as this should fail the same on either version. Could you confirm if that execution also had multiple nodes? Were there any other changes between the working and non working charts?

**Comment by JonnyBDev** [comment]
Well, I think I have to revert that statement - at least half. Had a setup running that night with around 500 restart from qryn and somehow it worked. Tried to replicate it like written above and it's not working.

**Comment by lmangani** [comment]
@JonnyBDev I appreciate the update! it confirms the theory of our lead developer @akvlad . We'll push a change to support this type of deployment in the coming days to get you going. This feedback about the variety of possible setups out there is useful to keep the current qryn version well until the next is released. I appreciate your time!

**Comment by JonnyBDev** [comment]
You're welcome. If you need anything from me to test hit me up.

**Comment by JonnyBDev** [comment]
fyi. Had the logs still open for my deployment from 30m ago (22.9.5 CH + 2.3.12 Qryn). Somehow the logs are giving me some more information and it seems, that it advanced in the setup - somehow? At least there are more successfully looking logs than earlier.

```
> qryn@2.3.12 start
> node qryn.js

{"level":30,"time":1695815859258,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"Initializing DB... cloki"}
{"level":30,"time":1695815859312,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"xxh ready"}
{"level":30,"time":1695815859510,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"v1 -> v2"}
{"level":20,"time":1695815859510,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"CREATE TABLE IF NOT EXISTS samples_v3\n    (\n      fingerprint UInt64,\n      timestamp_ns Int64 CODEC(DoubleDelta),\n      value Float64 CODEC(Gorilla),\n      string String\n    ) ENGINE = MergeTree\n    PARTITION BY toStartOfDay(toDateTime(timestamp_ns / 1000000000))\n    ORDER BY (timestamp_ns)"}
{"level":30,"time":1695815859581,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"xxh ready"}
{"level":30,"time":1695815859613,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"v2 -> v3"}
{"level":20,"time":1695815859613,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"CREATE TABLE IF NOT EXISTS settings\n    (fingerprint UInt64, type String, name String, value String, inserted_at DateTime64(9, 'UTC'))\n    ENGINE = ReplacingMergeTree(inserted_at) ORDER BY fingerprint"}
{"level":30,"time":1695815859672,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"v3 -> v4"}
{"level":20,"time":1695815859672,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"DROP TABLE IF EXISTS samples_read"}
{"level":30,"time":1695815859751,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"v4 -> v5"}
{"level":20,"time":1695815859751,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"CREATE TABLE IF NOT EXISTS samples_read\n   (fingerprint UInt64,timestamp_ms Int64,value Float64,string String)\n   ENGINE=Merge('cloki', '^(samples|samples_v2)$')"}
{"level":30,"time":1695815859795,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"v5 -> v6"}
{"level":20,"time":1695815859796,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","msg":"CREATE VIEW IF NOT EXISTS samples_read_v2_1 AS \n    SELECT fingerprint, timestamp_ms * 1000000 as timestamp_ns, value, string FROM samples_read"}
{"level":50,"time":1695815859816,"pid":19,"hostname":"qryn-654c6c5db7-78tdc","name":"qryn","err":"Error: Request failed with status code 404\nResponse: [404] undefined\nError: Request failed with status code 404\n    at createError (/app/node_modules/axios/lib/core/createError.js:16:15)\n    at settle (/app/node_modules/axios/lib/core/settle.js:17:12)\n    at IncomingMessage.handleStreamEnd (/app/node_modules/axios/lib/adapters/http.js:269:11)\n    at IncomingMessage.emit (node:events:525:35)\n    at endReadableNT (node:internal/streams/readable:1358:12)\n    at processTicksAndRejections (node:internal/process/task_queues:83:21)","msg":"Error starting qryn"}
npm notice 
npm notice New major version of npm available! 8.19.4 -> 10.1.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v10.1.0>
npm notice Run `npm install -g npm@10.1.0` to update!
npm notice
```

**Comment by lmangani** [comment]
@JonnyBDev Changes are ready. Would you be able to validate this branch? https://github.com/metrico/qryn/pull/350

**Comment by JonnyBDev** [comment]
Thanks for the changes. Will check in the next hour.

**Comment by JonnyBDev** [comment]
@lmangani Can confirm that Qryn works with the latest Clickhouse Chart (https://artifacthub.io/packages/helm/bitnami/clickhouse/4.0.4) on Kubernetes v1.27.1. No errors in the initialization of the deployment.

**Comment by lmangani** [comment]
Thanks @JonnyBDev for the configuration and to @akvlad for taking the time to make this fix over the weekend!

**Comment by lmangani** [comment]
@JonnyBDev changes are now merged to master and released into qryn 2.4.x thanks for your assistance!
